### ISSUE TITLE
Incremental compilation of java modules is broken with Gradle 7.6 

### ISSUE THREAD

**Issue by breskeby** [gradle/gradle#23067]
After updating our elasticsearch build to Gradle 7.6 we are seeing weird compilation issues when using incremental compilation.
The only two workarounds are 
1. disable incremental compilation
2. run clean task solves the problem

### Expected Behavior

It should not be required to run the clean task for correct java module compilation

### Current Behavior

When changing the sources slightly we see java module related error messages during `compileJava` like

```
package org.elasticsearch.core is not visible
```
 
### Context

We seen this when updating the elasticsearch build to from 7.5.1 to Gradle 7.6 

### Steps to Reproduce 

- git clone https://github.com/breskeby/elasticsearch.git
- git checkout repro-incremental-compile-issue
- git checkout b0a82469130813d4f6894003f6bfab528d37e68a
- ./gradlew :server:classes
- git checkout 04c6269a2f74afe61774a9711315484d70fb63eb
- ./gradlew :server:classes
  
### Your Environment

Build scan URL: https://gradle.com/s/amjv2v4a4qmym

**Comment by breskeby** [comment]
Reproducing the expected behaviour with Gradle 7.5.1 can be done as follows:

```
git clone https://github.com/breskeby/elasticsearch.git
git checkout repro-incremental-compile-issue-751
git checkout 5adb055cc110d945e9221cbf452f19fffa07a16d
./gradlew clean :server:classes
git checkout f9f9f5b22e957760510c81116790c26575120095
./gradlew :server:classes```

**Comment by asodja** [comment]
Moved to 8.0RC1, since we have to finish 8 in next weeks. But we will backport the fix to 7.6.1.

**Comment by asodja** [comment]
Note: If the fix still causes some issues for you, you can use in future Gradle 8 (and future Gradle 7.6.1):
```
tasks.withType(JavaCompile).configureEach {
  options.incrementalAfterFailure = false
}
```
that should make compilation behavior the same as with Gradle 7.5.
